refactor(sonar): switch to Sonar scan in GO

CCFenner opened the pull request:
**changes:**
- use `piper` bin in sonarExecuteScan
- encorporate #1341
- fix type of `options` parameter (`String` -> `List`)

redehnroV commented on cmd/sonarExecuteScan.go:
```diff
@@ -67,7 +67,7 @@ func runSonar(options sonarExecuteScanOptions, client piperhttp.Downloader, runn
 	}
 	if len(options.ProjectVersion) > 0 {
-		sonar.addOption("sonar.projectVersion=" + options.ProjectVersion)
+		sonar.addOption("sonar.projectVersion=" + handleArtifactVersion(options.ProjectVersion))
```
maybe a comment where this function is defined?

stippi2 commented on cmd/sonarExecuteScan.go:
```diff
@@ -127,6 +132,10 @@ func handlePullRequest(options sonarExecuteScanOptions) error {
 			sonar.addOption("sonar.pullrequest.branch=" + options.ChangeBranch)
 			sonar.addOption("sonar.pullrequest.provider=" + provider)
 		}
+	} else {
+		if len(options.BranchName) > 0 {
```
Minor: Since you already check an option on the outer if level, you could combine this into an "else if", no?

stippi2 commented on resources/metadata/sonar.yaml:
```diff
@@ -5,6 +5,14 @@ metadata:
 spec:
   inputs:
     params:
+      - name: instance
```
Where is `instance` used? I don't see it added in the options/config handling.

CCFenner replied:
This was the best way to use parameters in groovy without changing the whole `contextConfig` resolution.

stippi2 commented on vars/sonarExecuteScan.groovy:
```diff
@@ -1,231 +1,88 @@
-                            config.options.add("sonar.pullrequest.github.repository=${config.githubOrg}/${config.githubRepo}")
-                            break
-                        default: error "Pull-Request provider '${config.pullRequestProvider}' is not supported!"
+                withSonarQubeEnv(stepConfig.instance) {
```
Ah, so this is where "instance" is used.

stippi2 commented on vars/sonarExecuteScan.groovy:
```diff
@@ -1,231 +1,88 @@
+            "PIPER_parametersJSON=${groovy.json.JsonOutput.toJson(stepParameters)}",
+        ]) {
+            // get context configuration
+            Map config = readJSON(text: sh(returnStdout: true, script: "./piper getConfig --contextConfig --stepMetadata '.pipeline/tmp/${METADATA_FILE}'"))
+            echo "Config: ${config}"
+
+            Map stepConfig = readJSON(text: sh(returnStdout: true, script: "./piper getConfig --stepMetadata '.pipeline/tmp/${METADATA_FILE}'"))
+            echo "StepConfig: ${stepConfig}"
```
This will not use custom defaultConfigs or a custom config file. See my recent changes to piperExecuteBin, already on master. The code isn't easily re-used, though, maybe I should have added the two methods to PiperGoUtils.

CCFenner replied:
#1346

OliverNocon replied:
I think we should rather tackle the default topic properly via the binary instead of fiddling around with parameters.

From my POV it is okay to not adapt the recent change here and rather rely on a proper solution in the binary default resolution to be done soon.

stippi2 replied:
Sure, I just wanted to point it out.

CCFenner replied:
I will implement the change to support *custom defaults* and *project configuration* for now. But I do not favor the approach to do this is Groovy.

stippi2 commented on vars/sonarExecuteScan.groovy:
```diff
@@ -1,231 +1,88 @@
-                    config.options.add("sonar.login=$SONAR_TOKEN")
-                    workerForSonarAuth(config)
+            // load certificates into cacerts file
+            loadCertificates([customTlsCertificateLinks: stepConfig.customTlsCertificateLinks, verbose: stepConfig.verbose])
```
Not 100% sure, but I think you can omit the square brackets: Groovy will convert the key:value pairs to a Map automatically.

stippi2 commented on vars/sonarExecuteScan.groovy:
```diff
@@ -1,231 +1,88 @@
-                        default: error "Pull-Request provider '${config.pullRequestProvider}' is not supported!"
+                withSonarQubeEnv(stepConfig.instance) {
+                    withCredentials(credentials) {
+                        sh "./piper ${STEP_NAME}"
```
Here you don't use custom defaults or config file location either. Maybe you could use --stepConfig and pass the already resolved stepConfig map as JSON. But I am not sure if that will then ignore the credentials.

stippi2 commented on vars/sonarExecuteScan.groovy:
```diff
@@ -1,231 +1,94 @@
+                        credentials.add(string(credentialsId: config.githubTokenCredentialsId, variable: 'PIPER_githubToken'))
                 }
+                environment.add("PIPER_changeId=${env.CHANGE_ID}")
+                environment.add("PIPER_changeBranch=${env.CHANGE_TARGET}")
+                environment.add("PIPER_changeTarget=${env.CHANGE_BRANCH}")
```
It looks like you mixed these two up.

CCFenner replied:
```suggestion
                environment.add("PIPER_changeBranch=${env.CHANGE_BRANCH}")
                environment.add("PIPER_changeTarget=${env.CHANGE_TARGET }")
```

daniel-kurzynski commented:
I just tested it in on of our Cloud SDK pipelines and it failed:
```
[2020-04-07T11:52:11.343Z] time="2020-04-07T13:52:11+02:00" level=info msg="ERROR: Error during SonarQube Scanner execution" library= stepName=sonarExecuteScan

[2020-04-07T11:52:11.343Z] time="2020-04-07T13:52:11+02:00" level=info msg="ERROR: You must define the following mandatory properties for 'Unknown': sonar.projectKey" library= stepName=sonarExecuteScan

[2020-04-07T11:52:11.343Z] time="2020-04-07T13:52:11+02:00" level=info msg="ERROR: " library= stepName=sonarExecuteScan

[2020-04-07T11:52:11.343Z] time="2020-04-07T13:52:11+02:00" level=info msg="ERROR: Re-run SonarQube Scanner using the -X switch to enable full debug logging." library= stepName=sonarExecuteScan

[2020-04-07T11:52:11.343Z] time="2020-04-07T13:52:11+02:00" level=fatal msg="Execution failed" error="running command 'sonar-scanner' failed: cmd.Run() failed: exit status 1" library= stepName=sonarExecuteScan
```

with 
```
StepConfig: [githubApiUrl:https://api.github.com, instance:*********, options:[sonar.java.binaries=application/target/classes/, sonar.coverage.exclusions=**.js,unit-tests/**,integration-tests/**,performance-tests/**,**.xml,**/target/**, sonar.jacoco.reportPaths=integration-tests/target/coverage-reports/jacoco.exec,unit-tests/target/coverage-reports/jacoco.exec, sonar.java.libraries=./s4hana_pipeline/maven_local_repo/**, sonar.jacoco.reportPaths=s4hana_pipeline/reports/coverage-reports/unit-tests.exec,s4hana_pipeline/reports/coverage-reports/integration-tests.exec, sonar.sources=./application, sonar.projectKey=book-address-manager], projectVersion:1.0-2020-04-07T114249UTC_******, pullRequestProvider:GitHub, sonarScannerDownloadUrl:https://binaries.sonarsource.com/Distribution/sonar-scanner-cli/sonar-scanner-cli-4.3.0.2102-linux.zip]
```

CCFenner commented:
The sonar options are missing the `-D` prefix.

daniel-kurzynski commented:
But this means it's a breaking change? Because with the old step the same code was working!?

CCFenner commented:
The issue is, that the `options` parameter can be a list in groovy but in go it's a string. 

One option would be to read the step config in groovy (which is anyhow necessary in this step), prefix the options and hand them in explicitly.

daniel-kurzynski commented:
>The issue is, that the options parameter can be a list in groovy but in go it's a string.

But we have also cases where we allow the configuration to be a list:
https://github.com/SAP/jenkins-library/blob/master/resources/metadata/mavenExecute.yaml#L23

Why is that not the case for sonar?
https://github.com/SAP/jenkins-library/blob/master/resources/metadata/sonar.yaml#L58

CCFenner commented:
> Why is that not the case for sonar?
> https://github.com/SAP/jenkins-library/blob/master/resources/metadata/sonar.yaml#L58

Fixed, changed parameter to be a list.

daniel-kurzynski commented:
Still failing because of the same issue. I think the `-D` is still missing which was added in groovy: `https://github.com/SAP/jenkins-library/blob/master/vars/sonarExecuteScan.groovy#L143`

daniel-kurzynski commented on cmd/sonarExecuteScan_test.go:
```diff
@@ -107,7 +107,7 @@ func TestRunSonar(t *testing.T) {
 		}
 		options := sonarExecuteScanOptions{
-			Options: "-Dsonar.projectKey=piper",
+			Options: []string{"-Dsonar.projectKey=piper"},
```
```suggestion
			Options: []string{"sonar.projectKey=piper"},
```

daniel-kurzynski commented:
Works now. Thanks!

CCFenner commented:
/it

CCFenner commented:
@daniel-kurzynski can you try to trigger IT tests, seems I'm not allowed to do..

CCFenner commented:
/it
